subprojects/docs/src/docs/userguide/tasks.xml

```diff
@@ -290,52 +290,673 @@
+                                evaluation of the property values.
+                            </td>
+                        </tr>
+                    </tfoot>
+                </table>
+
+                <para>The last two annotations in the table are a special case as they don’t declare
+                    either task inputs or task outputs. So why use them? Because you can make use of
+                    the Java Gradle Plugin Development plugin to help you develop and publish your own
+                    plugins. And this plugin will check whether any properties of your custom task
+                    classes lack an annotation. This protects you against forgetting to add an appropriate
+                    annotation during development.
+                </para>
+
+                <para>Custom task classes are an easy way to bring your own build logic into the arena
+                    of incremental build, but you don’t always have that option. Perhaps you are working
+                    with a task whose source you can’t modify or with an ad-hoc task. In either case,
+                    you can still take advantage of incremental build by using the <literal>TaskInputs</literal>
+                    and <literal>TaskOutputs</literal> runtime APIs.
+                </para>
+            </section>
+
+            <section id="sec:task_input_output_runtime_api">
+                <title>Runtime API</title>
+                <para>When you don’t have access to the source for a custom task class, there is no way
+                    to add any of the annotations we covered in the previous section. Fortunately, Gradle
+                    provides a runtime API that provides access to the same functionality.
+                </para>
+
+                <section id="sec:runtime_api_for_adhoc">
+                    <title>Using it for ad-hoc tasks</title>
+                    <para>This runtime API is provided through a couple of aptly named properties that
+                        are available on every Gradle task:</para>
+                    <itemizedlist>
+                        <listitem>
+                            <literal>inputs</literal> of type <apilink class="org.gradle.api.tasks.TaskInputs"/>
+                        </listitem>
+                        <listitem>
+                            <literal>outputs</literal> of type <apilink class="org.gradle.api.tasks.TaskOutputs"/>
+                        </listitem>
+                    </itemizedlist>
+                    <para>These objects have methods that allow you to specify files, directories and
+                        values which constitute the task’s inputs and outputs. In fact, the runtime API
+                        has almost feature parity with the annotations. All it lacks is validation of
+                        whether declared files are actually files and declared directories are directories.
```

Comment:
Output directories [are created](https://github.com/gradle/gradle/blob/1ca45860f7df7502d15056a26c10ea60749266ea/subprojects/core/src/main/java/org/gradle/api/internal/project/taskfactory/OutputDirectoryPropertyAnnotationHandler.java#L54-L54) when registered via the output annotations, but not via `TaskOutputs`.

Input files and directories are checked to be the right type (and whether or not they exist) when registered via the annotations, but not via `TaskInputs`.

Future goal is to implement full parity between the two APIs. I guess there is no real reason for them not to have it today.